### PR TITLE
Add a method to access default binary messenger from test bindings

### PR DESCRIPTION
**Pull request by nturgut**
Add an alternative binary messenger to TestWidgetsFlutterBinding to be used by integration tests

## Description

TestWidgetsFlutterBinding only uses test platform messages: TestDefaultBinaryMessenger

This way, in the integration tests  such as e2e tests we can bind a real binary messenger rather than a fake one if we want to.

For a detailed usage see demo test: https://github.com/flutter/flutter/pull/49833/files

## Related Issues

Related to running integration tests for web using e2e: https://github.com/flutter/flutter/issues/50212

## Tests

The change is in the test bindgings.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by nturgut** [comment]
> It looks like this pull request may not have tests. Please make sure to add tests before merging. If you need an exemption to this rule, contact Hixie.
> 
> **Reviewers**: Read the [Tree Hygiene page](https://github.com/flutter/flutter/wiki/Tree-hygiene#how-to-review-code) and make sure this patch meets those guidelines before LGTMing.

The change is in the test bindings.

**Comment by collinjackson** [comment]
The `TestDefaultBinaryMessenger` exposes the wrapped `BinaryMessenger` as a public `delegate` property, so it seems like you wouldn't need to change the API.

See discussion in this thread on the original PR: https://github.com/flutter/flutter/pull/49833/files#r382864252

File: packages/flutter_test/lib/src/binding.dart [hunk 1]
```diff
@@ -275,6 +275,17 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
...
   }
 
+  /// For using the default implementaion of [BinaryMessenger] instead of
+  /// test version.
+  ///
+  /// This method is useful for integration tests which uses real platform
+  /// messages such as e2e.
+  ///
+  /// See: https://pub.dev/packages/e2e
+  BinaryMessenger createRealBinaryMessenger() {
```
Comment (yjbanov) [inline packages/flutter_test/lib/src/binding.dart]: Is this the right binding for e2e testing? Should we use something that doesn't mock stuff out? For example, this one also mocks out network. What do you think of having a specialized `E2eWidgetsFlutterBinding` (provided by `package:e2e`) that provides the exact implementation of the binary messenger it needs. See, for example, [_DriverBinding](https://github.com/flutter/flutter/blob/bf8e2c1449423a6f2e00919f8ee329f742390cd9/packages/flutter_driver/lib/src/extension/extension.dart#L45) provided by `package:flutter_driver` for driver tests.

/cc @Hixie who IIRC designed these bindings.

Comment (nturgut) [reply packages/flutter_test/lib/src/binding.dart]: e2e is on top of LiveTestWidgetsFlutterBinding. So there is one more parent in between.

Comment (yjbanov) [reply packages/flutter_test/lib/src/binding.dart]: Ah, at the time of writing my comment I didn't know that you were already adding `E2eWidgetsFlutterBinding`. I guess the issue might be that `TestFlutterWidgetsBinding` does more than just support `WidgetTester`. Perhaps we should reduce the scope of `TestFlutterWidgetsBinding` to not provide opinion beyond what `WidgetTester` needs, and have its subclasses add those opinions. For example, `TestFlutterWidgetsBinding` shouldn't override `createBinaryMessenger` and shouldn't mock out `HttpClient`. Then `E2eWidgetsFlutterBinding`, which is designed to run in a real non-mock environment, could use real implementations of those things. `flutter test`, OTOH, would mock stuff out more aggressively.

**Comment by nturgut** [comment]
Closing the PR, since we decided to change the inheritance chain of E2E bindings.

**Pull request closed without merging**